feat(yield): display cow amm banner conditionally

shoom3301 opened the pull request:
# Summary

Bound the banner displaying logic to "real" data (which is actually still mocked).

<img width="506" alt="image" src="https://github.com/user-attachments/assets/860c5914-1795-4d72-96f3-a1e0d4dfffe0">

# To Test

Should satisfy [figma logic](https://www.figma.com/design/iIblVIbidg7oG3SdgNCsde/Vampire-Attack-2024-Sept?node-id=0-1&node-type=canvas&t=2w189zgODUthx0Pj-0)

There are three main cases:
### 1. If a user has some LP-tokens and there is an alternative CoW AMM pools and one of pools has higher APY than opposite.

<img width="502" alt="image" src="https://github.com/user-attachments/assets/58df1221-93b2-4ce6-bb32-26d4a56ddda0">

> In the blue section we always display a pool with a highest APY
> But in "One-click convert" section we display icons of all providers where user has liquidity

### 2. If there is average APY information and CoW AMM average is better than others.
The info is presented like:
```
  [LpTokenProvider.COW_AMM]: { apy: 7.3, },
  [LpTokenProvider.UNIV2]: { apy: 3.1, },
  [LpTokenProvider.CURVE]: { apy: 0.4, },
  [LpTokenProvider.PANCAKE]: { apy: 0.2, },
  [LpTokenProvider.SUSHI]: { apy: 0.41, },
```

We calculate average of all presented providers: (3.1+0.4+0.2+0.41)/4=1.0275
And caculate diff = 7.3-1.0275=6.2725
And if the diff is > 0 we display the banner:

<img width="506" alt="image" src="https://github.com/user-attachments/assets/803965ae-98b2-4093-a517-48ba0ad93bac">

> There might me from one up to four displayed providers

### 3. When two cases above are not working we display:

<img width="506" alt="image" src="https://github.com/user-attachments/assets/7564109c-4be7-4fb7-b8bf-6ac13fd6fce7">

## Mocks overrides

For testing purposes I've added localStorage overrides:

To reproduce first scenario:

```
localStorage.setItem('MOCK_POOL_INFO', JSON.stringify({
      // Sushi AAVE/WETH
      '0xd75ea151a61d06868e31f8988d28dfe5e9df57b4': {
        apy: 1.89,
        tvl: 157057,
        feeTier: 0.3,
        volume24h: 31.19,
      },
      // CoW AMM AAVE/WETH
      '0xf706c50513446d709f08d3e5126cd74fb6bfda19': {
        apy: 30.07,
        tvl: 52972,
        feeTier: 0.3,
        volume24h: 10,
      },
    }))
```

To reproduce second scenario:

```
localStorage.setItem('POOLS_AVERAGE_DATA_MOCK', JSON.stringify({
      COW_AMM: { apy: 30.3, },
      UNIV2: { apy: 3.1, },
      CURVE: { apy: 0.4, },
      PANCAKE: { apy: 0.2, },
      SUSHI: { apy: 0.41, },
      }))
```

> Execute the code in console and reload the page
> Don't forget to remove the localStorage keys after testing!

elena-zh commented:
Hey @shoom3301 , great!
But I have not been able to reproduce the 1st scenario. Each time after the page refresh I've got this banner:
![image](https://github.com/user-attachments/assets/6e5b592f-4f73-4abc-a091-d4996acaf5dc)


Other issues that I've faced: 
1. When an account does not have LP tokens, a bit different message should be displayed (see it in #4981) 
![no lp tokens](https://github.com/user-attachments/assets/e643949e-7595-4f68-96b7-c9840de721ba)
2. When there are no CoW AMM pools in the selected chain, it would be also nice to change the banner to the one mentioned above. This case is related to Sepolia, but in the near future we can face this issue on Base.
![no pools](https://github.com/user-attachments/assets/9b6e4af3-9abc-4269-a171-2b49b20d6af9)
3. I think it would be nice to improve the blue area layout when there are 304 pools are mentioned: percentage takes a lot of room, but the text area is too small (cc @fairlighteth )
![reorganise](https://github.com/user-attachments/assets/15492a95-f527-44db-98e3-acc49838069e)
4. Could you please double check how the average difference is calculated? See: I have 1 external pool. And the difference should be 2.59, but the banner shows 2.79%. Or I don't get something? 
![2 59](https://github.com/user-attachments/assets/9c0863aa-bd84-49f9-92d3-f8570e07af8e)


Thanks!

alfetopito commented on libs/tokens/src/state/tokenLists/tokenListsStateAtom.ts:
```diff
@@ -56,7 +56,7 @@ export const allListsSourcesAtom = atom((get) => {
 // Lists states
 export const listsStatesByChainAtom = atomWithStorage<TokenListsByChainState>(
-  'allTokenListsInfoAtom:v3',
+  'allTokenListsInfoAtom:v4',
```
Migration? Otherwise users will lose previously manually added tokens lists

shoom3301 replied:
No need to add a migration, this state just a cache.
This store is filled by `TokensListsUpdater`.
User-added lists are stored in `userAddedListsSourcesAtom`.

shoom3301 commented:
@elena-zh 

>But I have not been able to reproduce the 1st scenario. Each time after the page refresh I've got this banner:

For the case you also need to use an account with some LP balance. Please try 0x0b80f483c3f842cc682ff22408b36b46aa7e9e74 on Mainnet.

>When an account does not have LP tokens, a bit different message should be displayed

This state should be shown **only** when CoW AMM average apy is better than others, see POOLS_AVERAGE_DATA_MOCK

>When there are no CoW AMM pools in the selected chain, it would be also nice to change the banner to the one mentioned above. This case is related to Sepolia, but in the near future we can face this issue on Base.

Fixed. Now the banner will be only shown if there are some CoW AMM pools.

>I think it would be nice to improve the blue area layout when there are 304 pools are mentioned: percentage takes a lot of room, but the text area is too small 

Adjusted font size

>Could you please double check how the average difference is calculated? See: I have 1 external pool. And the difference should be 2.59, but the banner shows 2.79%. Or I don't get something?

Thank you! I've been taking cow amm into account, but I should not. Fixed

shoom3301 commented:
@elena-zh I also improved the banner displaying.
Now it displays only when LP balances are loaded.
Details: https://cowservices.slack.com/archives/C0361CDG8GP/p1730203278660829?thread_ts=1730137351.410359&cid=C0361CDG8GP

elena-zh commented:
Hey @shoom3301 , thank you for explanation :)

Some issues: 
1. When change an account/wallet, I can see that 5-7 seconds the previous banner is displayed, then is changed to the new one. IMO, it would be nice to hide it and show a new one with updated data
2. UI: there are some discrepancies in styles for average case: 
 - tokens should not have a dark border
 - main banner font-size should be bigger
actual: 
![image](https://github.com/user-attachments/assets/debd85bc-f4fa-4621-b8a7-1bc3d15b922e)

design: 
![text and border](https://github.com/user-attachments/assets/240d699c-d524-425c-b0f8-c64771041fd6)

3. APR does not look good in the list: 
![apr](https://github.com/user-attachments/assets/96034ada-332f-4600-9d66-e6bba8adb707)


**Questions to discuss:** 
1. It would be nice to get a clear understanding which option should be displayed in the banner when there are no LP tokens for an account
![pick a version (no lp tokens)](https://github.com/user-attachments/assets/ca88459b-49be-4db5-819b-8e6b5b94da9a)
2. I think, some scenarios are missing here: 
2.1. one CoW AMM pool against 1 pool in Uni (sushi/pancake/etc.), lower yield in CoW APP pool. According to scenario implemented in #4981, this message should be displayed: 
![missing scenario](https://github.com/user-attachments/assets/aeba85a5-314e-445b-a3a8-963622a8e208)
but we show this banner: 
![what we show](https://github.com/user-attachments/assets/a9011b67-560f-4db7-a692-7f5a2ee749b2)
2.2. there are at least 2 pools in uni (sushi/pancake/etc.) against 2 pools in CoW AMM. What banner should be displayed/token should be picked when
- 2 CoW AMM pools have higher yield -->  I think, scenario 1, but what token should be picked when click on the 'Booost' button?
- 2 pools on other sources have higher yield --> I think, scenario 3, but what token should be picked when click on the 'Booost' button?
- 1 CoW AMM pool has a higher yield, the 2nd one is lower --? 

3. I'm connected to an account, close all banners (main and the one in the token list), then I switch to another account --> the banner does not appear again. Is this an expected behavior? Or we should show at least one banner (f.e. in the token list) per account? 

4. Same question is related to the chain. When I'm connected to GC, for example, and the banner/data is loaded per this chain. I close the banner, and it will never appear again for mainnet, even if a user has some LPs there. Correct?
5. When we say 'higher APR' for a certain pool, we show the APR of a CoW AMM pool. Not a difference between CoW AMM pool and the one in another source. Is this an expected behavior? 
~~6. (resolved question: see this [thread](https://cowservices.slack.com/archives/C07M5DZHMHC/p1730210618621329?thread_ts=1730208702.807899&cid=C07M5DZHMHC)) Banner should not appear in a chain when no CoW AMM pool created.~~

Thanks

shoom3301 commented:
@elena-zh thanks!
1. Fixed
2. tokens should not have a dark border - fixed, main banner font-size should be bigger - font-size of this text is dynamic, +30.1% is bigger that +2.5% so the right part is adjusted
3. Fixed

elena-zh commented:
Hey @shoom3301 , thanks!

As for the UI fixes: 
1. There APY instead of APR in a pool details: 
![apr](https://github.com/user-attachments/assets/e81dd1a9-46f0-4f91-b142-82dfb1ecadf5)
![apy](https://github.com/user-attachments/assets/22a6a1ee-9d46-42ef-8a3c-3de365f843b7)
2. After a wallet connection, there are different banners appear for an account. See the [video](https://www.loom.com/share/c3c95976ea0349bc928e716e45b8ec0a): the banner's message is changed when navigate to the Yield tab.

shoom3301 commented:
@elena-zh 
1. Fixed
2. Fixed in the next PR https://github.com/cowprotocol/cowswap/pull/5046

elena-zh commented:
Hey @shoom3301 , for now, changes LGTM. 
However, there are still some pending questions like: 
**Questions to discuss:** 
1. It would be nice to get a clear understanding which option should be displayed in the banner when there are no LP tokens for an account
![pick a version (no lp tokens)](https://github.com/user-attachments/assets/ca88459b-49be-4db5-819b-8e6b5b94da9a)
2. What should we display in the banner and which tokens to pick when there are
- 1 CoW AMM pool has a higher APY but a lower volume, and the 2nd one has a higher balance and a lower APY?

3. When we say 'higher APR' for a certain pool, we show the APR of a CoW AMM pool. Not a difference between CoW AMM pool and the one in another source. Is this an expected behavior?

shoom3301 commented:
@elena-zh 

>It would be nice to get a clear understanding which option should be displayed in the banner when there are no LP tokens for an account

In that case we go to 2nd cenario in the PR decription ( If there is average APY information and CoW AMM average is better than others.). Please call me if this description is not clear, I will try to explain live

>CoW AMM pool has a higher APY but a lower volume, and the 2nd one has a higher balance and a lower APY?

Currently we only check APY

>When we say 'higher APR' for a certain pool, we show the APR of a CoW AMM pool. Not a difference between CoW AMM pool and the one in another source. Is this an expected behavior?

No, I think it's a bug, I've fixed it in [this PR](https://github.com/cowprotocol/cowswap/pull/5046)

elena-zh reviewed (approved):
Approving this PR: questions can be discussed later and addressed  in a separate PR (if needed)

anxolin commented on apps/cowswap-frontend/src/common/containers/CoWAmmBanner/index.tsx:
```diff
@@ -21,18 +25,37 @@ interface BannerProps {
 
   const key = isTokenSelectorView ? 'tokenSelector' : 'global'
   const handleCTAClick = useCallback(() => {
+    const superiorAlternative = vampireAttackContext?.superiorAlternatives?.[0]
```
superior alternative? LoL
